icu4c/source/i18n/number_mapper.cpp

```diff
@@ -74,9 +74,11 @@ MacroProps NumberPropertyMapper::oldToNew(const DecimalFormatProperties& propert
             !properties.currencyPluralInfo.fPtr.isNull() ||
             !properties.currencyUsage.isNull() ||
             warehouse.affixProvider.get().hasCurrencySign());
-    CurrencyUnit currency = resolveCurrency(properties, locale, status);
-    UCurrencyUsage currencyUsage = properties.currencyUsage.getOrDefault(UCURR_USAGE_STANDARD);
+    CurrencyUnit currency;
+    UCurrencyUsage currencyUsage;
```

Comment:
I expanded the rest of this (long) function, searched for "currency", and found some code that uses these variables but is gated by conditions other than `useCurrency`.

- `if (!properties.currencyUsage.isNull())`
- `if (exportedProperties != nullptr)`

If we think we can guarantee that useCurrency is true there, then we should add `U_ASSERT(useCurrency)` there.
For the second one, maybe a default-constructed `currency` is fine if not `useCurrency`?
Otherwise, we may need to add a runtime check for useCurrency.

@sffc @mihnita @richgillam @eggrobin

Reply:
Markus is right; this is a concern.

For `if (!properties.currencyUsage.isNull())`, I wonder if this could just be changed to `useCurrency`?

For `exportedProperties->currency`, I agree that a default-constructed `currency` ought to be fine.

For `if (precision.fType == Precision::PrecisionType::RND_CURRENCY)`, I wonder if we could change this to `if (useCurrency && precision.fType == Precision::PrecisionType::RND_CURRENCY)`?

Reply:
Added an assertion to `if (!properties.currencyUsage.isNull())`, since useCurrency is true when `!properties.currencyUsage.isNull()`
Used `if (useCurrency && precision.fType == Precision::PrecisionType::RND_CURRENCY)`, thank you!